README.md
# t64fix v0.4.x

Bugfix branch of version 0.4

## New in version 0.4.0

- Add `--create` command: create a T64 image and write one or more PRG files to it.
- Have the option parser properly report a missing argument for string arguments,
  not just fail silently.
- Improve `make install`
- Add man page.


## Introduction

t64fix is a small tool to fix faulty T64 tape images. T64 images are container
files used by the old C64S emulator by Miha Peternel. The 'tape' in the name is
a misnomer as these files have nothing to do with actual tape dumps, they're
used as archives and for storing memory snapshots in C64S.

There are a lot of faulty t64 files out there, mostly created with tools which
were written by people who didn't really understand the format.


## Usage

Usage is relatively simple: `t64fix [OPTIONS] <SOURCE>`

When just checking if a t64 file is correct, just issue `t64fix <SOURCE>`. This
will verify the image and show its contents and any warnings on stdout. To fix
an image, issue `t64fix <SOURCE> -o <DESTINATION>`, this will verify and fix
\<SOURCE\> and write it to \<DESTINATION\>. Using the same for SOURCE and
DESTINATION is fine, SOURCE is read into memory and then closed, so using
`t64fix foo.t64 -o foo.t64` will fix foo.t64 and write it back. But obviously
the original file will be lost, so this is not advisable.


More 'advanced' use is available through a few command line switches:

| Option                                    | Description                                         |
|:----------------------------------------- |:----------------------------------------------------|
| `-q, --quiet`                             | don't output anything to stdout, for use in scripts |
| `-o, --output <fixed-image>`              | write fixed image to filesystem                     |
| `-e, --extract <index>`                   | extract file \<index\> from image                   |
| `-x, --extract-all`                       | extract all files, except memory snapshots          |
| `-c, --create <image> <list-of-files>`    | create t64 image and write on or more files to it   |
| `--help`                                  | show help                                           |
| `--version`                               | show version info                                   |


The `--quiet` option tells t64fix to not output any information on stdout, it
just returns an exit code (`EXIT_SUCCESS` or `EXIT_FAILURE`). See the bash
script `scripts/verify_multi.sh` for an example.



### Things that get verified and fixed

There are a few things that get verified and fixed:

#### Header magic string

There are a few different types of header magic strings out there, the correct
one being "C64S image file", padded to 32 bytes with 0x00. This is the string
found in t64 files generated by C64S 2.52.

#### Header tape version number

Version numbers seem to be either 0x100 or 0x101, no information on the
difference can be found, so this tool sets the version number to 0x101, which
is the version number C64S 2.52 uses.

#### Number of file records

Many t64 files have the number of available file records and the number of
used file records wrong, which can lead to tools not correctly reading files
from the container. If either of these numbers is zero, this tools adjusts that
number to one since t64 files with more than one record are rare, and seem to
have their record counts correct.

#### End addresses of files

The thing what makes tools and emulators choke on t64 files. Someone wrote a tool
called CONV64 which incorrectly writes the end address of any file as 0xc3c6.
This causes some tools or emulators to reject the file or silently ignore the
incorrect end address, which causes some decrunchers to fail.

This tool corrects the end address by sorting file records on their data offset
and then using that information, combined with the file size of the t64 file
to determine the correct end address.

#### C1541 file type

Another thing many images get wrong. File records are supposed to have a C1541
file type byte, which normally should be between 0x80 and 0x84. In reality this
is usually something like 0x01 or 0x44. This tool adjusts incorrect values to
0x82 (PRG), since t64 files can really only store PRG files (and C64S' FRZ files)


## Building and installing

To build t64fix simply run `make`. To install run `make install` as root.

Targets for `make`:

| Target          | Result
|:--------------- |:------------------------------------------------------------------------- |
| \<none\>        | Build t64fix without debugging enabled                                    |
| `all`           | Build t64fix without debugging enabled                                    |
| `clean`         | Remove t64fix and all intermediate objects, delete Doxygen documentation  |
| `debug`         | Build t64fix with debugging enabled                                       |
| `dist`          | Generate distribution tarball (`t64fix-$(VERSION).tar.gz`)                |
| `doc`           | Generate Doxygen documentation (in `doc/html`)                            |
| `install`       | Install t64fix executable and its man page                                |
| `install-bin`   | Install t64fix executable                                                 |
| `install-man`   | Install t64fix man page                                                   |
| `windist`       | Generate Windows distribution zipfile (`t64fix-win[32|64]-$(VERSION).zip` |



## Future

I probably won't be adding any more features to this tool, unless specifically
asked for them. Any bugs found I will fix, and of course accept patches for
them.

### TODO

- d64 support


### BUGS

none(?)


## Resources

- [t64fix GitHub project page](https://github.com/Compyx/t64fix/)
- [t64fix website](https://compyx.github.io/t64fix/)
- [T64 file format description](http://unusedino.de/ec64/technical/formats/t64.html)


CHANGES.md
# CHANGES

## A list of changes applied to the codebase, started too late as usual ;)

### 2021-09-01

* Add `debug` rule to Makefile, passes -DDEBUG via CPPFLAGS.
* Implement proper `install` rule in Makefile.
* Add `dist` rule to Makefile, creates a tarball of the sources. Excludes the reference documentation and the test files (t64, prg etc).
* Add man page.


### 2021-08-20

* Split t64.h into t64types.h and t64.h, with t64types.h containing the globally
  used defines and types.
* Properly convert PETSCII filename to ASCII in prg\_extract().
* Move command handling into separate functions.
* Remove implicit --output, no more `t64fix <org-file> <fixed-file>`.
* Remove fixed TODOs.


### 2021-08-19

* Properly translate generated tapename to PETSCII.
* Add base\_debug() macro.
* Pad filenames in directory with spaces ($20), not $00.
* Properly convert filenames from/to PETSCII.
* Remove filename padding when printing records to stdout.
* Translate tapename from PETSCII when printing to stdout.


### 2021-08-18

* Update Doxygen docblocks and the Doxyfile.
* Rework optparse to use xmalloc()-style allocation.
* Call abort() rather than exit() on allocation failures to allow debuggers to
  get useful information to work with.


### 2021-08-17

* Remove Windows-specific hacks for %zu and %zx format specifiers.
* Remove Makefile.win, the normal Makefile runs fine in MSys2.
* Remove `-g` command.
* Update README.
* Update copyright lines.


### 2021-08-15

* Finally finish the `--create` command.
* Bump version to 0.4.0


### 2020-07-18

* Update copyright to 2020


### 2018-06-06

* Add the -g/--groepaz command line switch.


### 2016-07-18

* Remove d64 support, too much work. Another project which does support d64
  and other formats will replace this project soon(ish)


### 2016-07-11

* Add extracting program files with -e/--extract \<index\>, or -x/--extract-all
  to extract all program files


### 2016-07-10

* Add simple command line parser (src/optparse.{c,h})
* Add quiet mode (-q/--quiet) for use in scripts


### 2016-07-08

* Add this file
* Start work on D64 support
* Apply patch by iAN CooG: print original magic header, not the fixed one



src/t64.c
/** \file   t64.c
 * \brief   T64 image handling
 *
 * \author  Bas Wassink <b.wassink@ziggo.nl>
 *
 * So far, all functions that fix stuff in the t64 files do so by writing into
 * t64_record_t and t64_image_t instances, so the actual t64 file isn't touched
 * until it is written to the OS, either using the original file or a copy.
 */

/*
t64fix - a small tool to correct T64 tape image files
Copyright (C) 2016-2021  Bas Wassink <b.wassink@ziggo.nl>

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <errno.h>

#include "base.h"
#include "cbmdos.h"
#include "petasc.h"

#include "t64.h"


/* {{{ Static data */

/** \brief  C1541 file types
 */
static const char *c1541_types[] = {
    "frz", "seq", "prg", "rel", "usr", "???", "???", "???"
};


/** \brief  The correct 'magic' for t64 files
 *
 * The string is in ASCII, padded to 32 bytes with 0x00. This is based on a
 * file generated by C64S 2.52.
 */
static const char c64s_magic[T64_HDR_MAGIC_LEN] = "C64S tape image file";


/** \brief  List of 'magic bytes' found in different t64 files
 */
static const char *magic_strings[] = {
    c64s_magic,         /* the correct one */
    "C64S tape file",
    "C64 tape image file"
};


/** \brief  Status strings for records
 */
static const char *status_strings[] = {  "OK", "fixed", "skipped" };


/* }}} */


/* {{{ T64 header functions */

/** \brief  Check magic bytes in header of \a image
 *
 * This functions checks the magic bytes in the header against known magic
 * bytes found in various tape images. The first entry checked is the correct
 * one, so if this function returns 0, we have the correct magic, if this
 * function returns > 0, we have some custom magic. If it returns < 0, no
 * magic was found.
 *
 * \param[in]   image   t64 image
 *
 * \return  index in `magic_strings` or -1 when not found
 */
static int t64_check_magic(const t64_image_t *image)
{
    size_t i;

    for (i = 0; i < sizeof magic_strings / sizeof magic_strings[0]; i++) {
        if (memcmp(magic_strings[i], image->data + T64_HDR_MAGIC, 
                    strlen(magic_strings[i])) == 0) {
            return (int)i;
        }
    }
    return -1;
}


/** \brief  Parse header to determine if it is a t64 image, apply some fixes
 *
 * This function checks the header's magic against known magic strings. If no
 * magic was found, 0 (false) is returned, otherwise, some header fixes are
 * applied and 1 (true) is returned
 *
 * \return  boolean
 */
static bool t64_parse_header(t64_image_t *image, int quiet)
{
    int result = t64_check_magic(image);
    if (result < 0) {
        if (!quiet) {
            printf("t64fix: fatal: couldn't find magic bytes, aborting\n");
        }
        return false;
    } else {
        if (result > 0) {
            if (!quiet) {
                printf("t64fix: warning: fixing header magic bytes\n");
            }
            image->fixes++;
        }
        strcpy((char *)(image->magic), magic_strings[result]);
    }
    /* get file record max and used counters */
    image->rec_max = get_uint16(image->data + T64_HDR_REC_MAX);
    image->rec_used = get_uint16(image->data + T64_HDR_REC_USED);
    if (image->rec_max == 0) {
        if (!quiet) {
            printf("t64fix: warning: maximum records count reported as "
                "0, adjusting to 1\n");
        }
        image->rec_max = 1;
        image->fixes++;
    }
    if (image->rec_used == 0) {
        if (!quiet) {
            printf("t64fix: warning: current records count reported as "
                "0, adjusting to 1\n");
        }
        /* this fix is required for the other fixes to work */
        image->rec_used = 1;
        image->fixes++;
    }

    if (image->rec_used > image->rec_max) {
        if (!quiet) {
            printf("t64fix: warning: header reports more used records than "
                    "available records (%d > %d), fixing\n",
                    (int)(image->rec_used), (int)(image->rec_max));
        }
        image->rec_used = image->rec_max;
        image->fixes++;
    }

    /* copy tape name: warning: not 0-terminated */
    memcpy(image->tapename, image->data + T64_HDR_NAME, T64_HDR_NAME_LEN);

    /* get tape version */
    image->version = get_uint16(image->data + T64_HDR_VERSION);
    return true;
}


/** \brief  Write correct header data in \a image
 *
 * \param[in]   image   t64 image
 */
static void t64_write_header(t64_image_t *image)
{
    /* write the correct magic */
    memcpy(image->data + T64_HDR_MAGIC, c64s_magic, T64_HDR_MAGIC_LEN);

    /* write tape name */
    memcpy(image->data + T64_HDR_NAME, image->tapename, T64_HDR_NAME_LEN);

    /* write version */
    set_uint16(image->data + T64_HDR_VERSION, 0x101);

    /* write number of available file records and number of used records */
    set_uint16(image->data + T64_HDR_REC_MAX, image->rec_max);
    set_uint16(image->data + T64_HDR_REC_USED, image->rec_used);
}

/* }}} */


/* {{{ T64 file record handling */

/** \brief  Comparison function for qsort() call to sort records on data offset
  *
  * \param[in]  p1  pointer to record 1
  * \param[in]  p2  pointer to record 2
  *
  * \return -1 if r1<r2, 0 if r1==r2, 1 if r1>r2
  */
static int compar_offset(const void *p1, const void *p2)
{
    const t64_record_t *r1 = p1;
    const t64_record_t *r2 = p2;

    if (r1->offset < r2->offset) {
        return -1;
    } else if (r1->offset > r2->offset) {
        return 1;
    } else {
        return 0;
    }
}


/** \brief  Comparison function for qsort() call to sort records on index
  *
  * \param[in]  p1  pointer to record 1
  * \param[in]  p2  pointer to record 2
  *
  * \return -1 if r1<r2, 0 if r1==r2, 1 if r1>r2
  */
static int compar_index(const void *p1, const void *p2)
{
    const t64_record_t *r1 = p1;
    const t64_record_t *r2 = p2;

    if (r1->index < r2->index) {
        return -1;
    } else if (r1->index > r2->index) {
        return 1;
    } else {
        return 0;
    }
}


/** \brief  Read a single record into \a record using \a data
 *
 * \param[in]   record  t64 file record
 * \param[in]   data   index in file records
 *
 * \return  new file record or NULL on malloc(3) failure
 */
static void t64_read_record(t64_record_t *record, const unsigned char *data)
{
    memcpy(record->filename, data + T64_REC_FILENAME, T64_REC_FILENAME_LEN);
    record->offset = get_uint32(data + T64_REC_CONTENTS);
    record->start_addr = get_uint16(data + T64_REC_START_ADDR);
    record->end_addr = get_uint16(data + T64_REC_END_ADDR);
    record->c64s_ftype = data[T64_REC_C64S_FILETYPE];
    record->c1541_ftype = data[T64_REC_C1541_FILETYPE];
    record->index = 0;
    record->status = T64_REC_OK;
}


/** \brief  Write corrected file record into image data
 *
 * \param[in]   record  t64 file record
 * \param[out]  dest    destination in image data
 */
static void t64_write_record(t64_record_t *record, unsigned char *dest)
{
    dest[T64_REC_C64S_FILETYPE] = record->c64s_ftype;
    dest[T64_REC_C1541_FILETYPE] = record->c1541_ftype;
    set_uint16(dest + T64_REC_START_ADDR, record->start_addr);
    set_uint16(dest + T64_REC_END_ADDR, record->real_end_addr);
    set_uint32(dest + T64_REC_CONTENTS, record->offset);
    memcpy(dest + T64_REC_FILENAME, record->filename, T64_REC_FILENAME_LEN);
}


/** \brief  Print file record in \a image at \a index on stdout
 *
 * \param[in]   image   t64 image
 * \param[in]   index   index in image
 */
static void t64_print_record(const t64_image_t *image, int index)
{
    t64_record_t *record = image->records + index;
    int size = record->end_addr - record->start_addr;
    char filename_asc[T64_REC_FILENAME_LEN + 1];
    int n;

    /* copy and translate PETSCII filename */
    filename_asc[T64_REC_FILENAME_LEN] = '\0';
    pet_to_asc_str(filename_asc, record->filename, T64_REC_FILENAME_LEN);
    /* strip padding */
    n = T64_REC_FILENAME_LEN - 1;
    while (n >= 0) {
        if (filename_asc[n] == 0x20) {
            filename_asc[n] = '\0';
        }
        n--;
    }

    /* print blocks and name */
    n = printf("%5d  \"%s\"",
                num_blocks((unsigned int)size),
                filename_asc);
    /* align filetype column */
    while (n < 26) {
        putchar(' ');
        n++;
    }
    printf("%s  $%04x-$%04x  $%04x-$%04x  %s\n",
            c1541_types[record->c1541_ftype & 0x07],
            record->start_addr, record->end_addr,
            record->start_addr, record->real_end_addr,
            status_strings[record->status]);
}
/* }}} */


/* {{{ T64 image file handling */

/** \brief  Allocate new, empty, t64 image
 *
 * \return  new t64 image
 */
static t64_image_t *t64_new(void)
{
    t64_image_t *image = base_malloc(sizeof *image);
    image->path = NULL;
    image->data = NULL;
    image->size = 0;
    image->records = NULL;
    image->rec_max = 0;
    image->rec_used = 0;
    image->fixes = 0;
    return image;
}


/** \brief  Open t64 container
 *
 * \param[in]   path    path to container
 * \param[in]   quiet   don't output anything on stdout/stderr
 *
 * \return  image or NULL on failure
 */
t64_image_t *t64_open(const char *path, int quiet)
{
    t64_image_t *image;
    long size;
    int i;

    image = t64_new();
    image->path = path;

    size = fread_alloc(&(image->data), path);
    if (size < 0) {
        /* error already reported by fread_alloc() */
        t64_free(image);
        return NULL;
    }
    /* store image size */
    image->size = (size_t)size;
    /* parse header for required information */
    if (!t64_parse_header(image, quiet)) {
        /* header parsing failed, bail: */
        t64_free(image);
        return NULL;
    }

    /* allocate and read records */
    image->records = base_malloc(sizeof *(image->records) * image->rec_used);
    for (i = 0; i < image->rec_used; i++) {
        t64_read_record(image->records + i,
                image->data + T64_RECORDS_OFFSET + i * T64_RECORD_SIZE);
        (image->records + i)->index= i;
    }

    return image;
}


/** \brief  Free memory used by t64 image
 *
 * \param[in]   image   t64 image
 */
void t64_free(t64_image_t *image)
{
    if (image->data != NULL) {
        base_free(image->data);
    }
    if (image->records != NULL) {
        base_free(image->records);
    }
    base_free(image);
}


/** \brief  Verify data in \a image, optionally fixing it
 *
 * \param[in]   image   t64 image
 * \param[in]   quiet   don't output anything on stdout/stderr
 *
 * \return  number of fixes, or -1 on error
 */
int t64_verify(t64_image_t *image, int quiet)
{
    size_t rec_size;    /* file size according to record */
    size_t act_size;    /* actual file size */
    int i;

    /* check maximum record count */
    if (image->rec_max == 0) {
        if (!quiet) {
            printf("t64fix: warning: maximum records count reported as "
               "0, adjusting to 1\n");
        }
        image->rec_max = 1;
        image->fixes++;
    }
    /* check number of used records */
    if (image->rec_used == 0) {
        if (!quiet) {
            printf("t64fix: warning: used records count reported as 0, "
                    "adjusting to 1\n");
        }
        /* this fix is required for the other fixes to work */
        image->rec_used = 1;
        image->fixes++;
    }

    /* Fix end addresses by sorting file records on data offset and then using
     * either the data offset of the next entry, or the length of the t64 file
     * to get the proper end address */

    /* sort entries based on offset */
    qsort(image->records, (size_t)(image->rec_used),
            sizeof *(image->records), compar_offset);

    /* process records, reporting any invalid data, optionally fixing it */
    for (i = 0; i < image->rec_used; i++) {
        t64_record_t *record = image->records + i;
        t64_record_t *next;

        /* check file type */
        if (record->c64s_ftype > 0x01) {
            /* memory snapshot */
            record->status = T64_REC_SKIPPED;
        } else {
            if (record->c1541_ftype < 0x80 || record->c1541_ftype >= 0x85) {
                if (!quiet) {
                    printf("t64fix: illegal value $%02x for C1541 file type, "
                            "assuming $82 (prg)\n", record->c1541_ftype);
                }
                record->c1541_ftype = 0x82;
                record->status = T64_REC_FIXED;
                image->fixes++;
            }

            record = image->records + i;
            if (i < image->rec_used - 1) {
                next = image->records + i + 1;
            } else {
                next = NULL;
            }


            /* get reported size */
            rec_size = (size_t)(record->end_addr - record->start_addr);
            /* determine actual size */
            if (i < image->rec_used - 1) {
                act_size = (size_t)(next->offset - record->offset);
            } else {
                act_size = (size_t)(image->size - record->offset);
            }
            if (rec_size != act_size) {
                if (i == image->rec_used -1 && rec_size < act_size) {
                    /* don't fix last record when actual size is larger: some
                     * T64's have padding for the last record */
                    continue;
                }
                if (!quiet) {
                    printf("t64fix: %d: reported size of $%04lx does not "
                            "match actual size of $%04lx\n",
                            i, (unsigned long)rec_size, (unsigned long)act_size);
                }
                record->status = T64_REC_FIXED;
                image->fixes++;
                record->real_end_addr = (unsigned short)(record->start_addr +
                        act_size);
            } else {
                record->real_end_addr = record->end_addr;
            }
        }

    }
    /* restore original record order */
    qsort(image->records, (size_t)(image->rec_used),
            sizeof *(image->records), compar_index);

    return image->fixes;
}


/** \brief  Print a 79 chars wide separator on stdout
 */
static void print_sep(void)
{
    puts("-------------------------------------------------------------------"
         "------------");
}


/** \brief  Dump t64 image information and file records on stdout
 *
 * \param[in]   image   t64 image
 */
void t64_dump(const t64_image_t *image)
{
    char tapename_asc[T64_HDR_NAME_LEN + 1];
    char magic[T64_HDR_MAGIC_LEN + 1];
    int i;

    /* copy tapename, translate PETSCII to ASCII */
    tapename_asc[T64_HDR_NAME_LEN] = '\0';  /* terminated name */
    pet_to_asc_str(tapename_asc, image->tapename, T64_HDR_NAME_LEN);
    /* remove padding */
    i = T64_HDR_NAME_LEN - 1;
    while (i >= 0 && tapename_asc[i] == 0x20) {
        tapename_asc[i--] = '\0';
    }

    /* copy magic */
    memcpy(magic, image->magic, T64_HDR_MAGIC_LEN);
    magic[T64_HDR_MAGIC_LEN] = '\0';    /* terminate magic */
    /* remove padding */
    i = T64_HDR_MAGIC_LEN - 1;
    while (i >= 0 && magic[i] == 0x20) {
        magic[i--] = '\0';
    }

    print_sep();
    printf("tape magic  : \"%s\"\n", magic);
    printf("tape version: %04x\n", image->version);
    printf("tape name   : \"%s\"\n", tapename_asc);
    printf("file records: %d/%d\n", (int)(image->rec_used),
            (int)(image->rec_max));
    print_sep();

    /* print file records */
    printf("blocks filename           type rep. memory  real memory  status\n");
    for (i = 0; i < image->rec_used; i++) {
        t64_print_record(image, i);
    }
    print_sep();
    if (image->fixes > 0) {
        printf("faulty image: fixes applied: %d\n", image->fixes);
    } else {
        puts("OK, proper image");
    }
}


/** \brief  Write t64 image to OS
 *
 * Write corrected image to host filesystem.
 *
 * This function stores corrected header and directory data in \a image before
 * writing to host.
 *
 * \param[in]   image   t64 image
 * \param[in]   path    path/filename of image
 *
 * \return  boolean
 */
bool t64_write(t64_image_t *image, const char *path)
{
    /* write corrected header in image data */
    t64_write_header(image);

    for (int i = 0; i < image->rec_used; i++) {
        t64_write_record(image->records + i,
                image->data + T64_RECORDS_OFFSET + (i * T64_RECORD_SIZE));
    }

    if (!fwrite_wrapper(path, image->data, image->size)) {
        return false;
    }

    return true;
}


/** \brief  Write PRG file into T64 image
 *
 * Write PRG file \a path into \a image and store its directory entry.
 *
 * \param[in]   image   t64 image
 * \param[in]   path    path to PRG file
 * \param[in]   index   index in directory of \a image
 * \param[in]   quiet   don't print anything on stdout
 *
 * \return  true if succesfully written
 */
static bool store_prg_file(t64_image_t *image,
                           const char *path,
                           int index,
                           bool quiet)
{
    t64_record_t record;
    uint8_t *data;
    uint8_t petname[CBMDOS_FILENAME_MAX];
    const char *ascname;
    const char *ext;
    long len;

    len = fread_alloc(&data, path);
    if (!quiet) {
        printf(".. file '%s' is %ld ($%04lx) bytes.\n",
               path, len, (unsigned long)len);
    }
    if (len < 0) {
        fprintf(stderr, "t64fix: failed to read file '%s'.\n", path);
        base_free(data);
        return false;
    }

    /* create directory record from file data */
    memset(&record, 0, sizeof(record));

    /*
     * set PETSCII filename using the file's ASCII basename
     */
    ascname = base_basename(path, &ext);
    asc_to_pet_str(petname, ascname, CBMDOS_FILENAME_MAX);
    /* pad filename with spaces */
    for (int i = CBMDOS_FILENAME_MAX - 1; i >=0; i--) {
        if (petname[i] == 0x00) {
            petname[i] = 0x20;
        }
    }
    memcpy(record.filename, petname, CBMDOS_FILENAME_MAX);

    /* set start, end and real_end */
    record.start_addr = get_uint16(data);
    record.end_addr = (uint16_t)(len - 2 + record.start_addr);
    record.real_end_addr = record.end_addr;
    if (!quiet) {
        printf(".... start address = $%04x\n", record.start_addr);
        printf(".... end address   = $%04x\n", record.end_addr);
    }

    /* set C64S file type */
    record.c64s_ftype = 0x01;   /* normal file */
    /* set CBMDOS file type and flags */
    record.c1541_ftype = CBMDOS_FILETYPE_PRG | CBMDOS_CLOSED_MASK;

    /* realloc data after header + dir */
    image->data = base_realloc(image->data, image->size + (size_t)len - 2);
    /* copy file data stripped of load address */
    memcpy(image->data + image->size, data + 2, (size_t)(len - 2));
    /* store data size */
    record.offset = (uint32_t)image->size;
    image->size += (size_t)len - 2;
    /* free file data */
    base_free(data);

    /* store directory entry in t64 image instance, not its raw data */
    memcpy(image->records + index, &record, sizeof(record));

    if (!quiet) {
        printf(".... added '%s'\n", ascname);
    }
    return true;
}


/** \brief  Create T64 image and add files
 *
 * \param[in]   path    name of T64 image to create
 * \param[in]   args    list of files to add
 * \param[in]   nargs   number of elements in \a args
 * \param[in]   quiet   don't output anything on stdout
 *
 * \return  new T64 image instance or `NULL` on error
 */
t64_image_t *t64_create(const char *path, const char **args, int nargs, bool quiet)
{
    t64_image_t *image;
    uint32_t dir_size;
    uint32_t data_offset;
    const char *img_name;
    const char *img_ext;
    size_t img_name_len;
    int index;

    /* create empty image */
    if (!quiet) {
        printf("Creating new t64 image '%s':\n", path);
    }
    image = t64_new();

    /* calculate directory size */
    dir_size = (unsigned int)nargs * T64_RECORD_SIZE;
    if (!quiet) {
        printf(".. directory size = $%04x\n", (unsigned int)dir_size);
    }
    /* calculate file data offset */
    data_offset = T64_RECORDS_OFFSET + dir_size;
    if (!quiet) {
        printf(".. data offset = $%04x\n", (unsigned int)data_offset);
    }

    /* allocate data for records */
    image->records = base_malloc(sizeof *(image->records) * (size_t)nargs);

    /* allocate data for header and directory and initialize header */
    image->data = base_malloc(data_offset);
    image->size = data_offset;
    memset(image->data, 0, data_offset);

    /* add files to image */
    for (index = 0; index < nargs; index++) {
        if (!store_prg_file(image, args[index], index, quiet)) {
            t64_free(image);
            return NULL;
        }
    }

    /* set directory size and entry count */
    image->rec_used = (uint16_t)index;
    image->rec_max = (uint16_t)index;

    /* set internal image name using the basename of the t64 file without
     * extension:
     */
    img_name = base_basename(path, &img_ext);

    /* T64's have their tape name padded with spaces, so clear the name by
     * writing spaces:
     */
    memset(image->tapename, 0x20, T64_HDR_NAME_LEN);
    base_debug("basename: '%s', extension: '%s'\n", img_name, img_ext);
    if (*img_ext == '\0') {
        /* no extension */
        img_name_len = strlen(img_name);
    } else {
        img_name_len = strlen(img_name) - strlen(img_ext) - 1;
    }
    base_debug("name len without ext = %zu\n", img_name_len);

    if (img_name_len > 0) {
        uint8_t tapename_pet[T64_HDR_NAME_LEN];

        /* copy at most 24 chars */
        if (img_name_len > T64_HDR_NAME_LEN) {
            img_name_len = T64_HDR_NAME_LEN;
        }
        /* translate to petscii */
        asc_to_pet_str(tapename_pet, img_name, T64_HDR_NAME_LEN);

        if (!quiet) {
            char tapename[T64_HDR_NAME_LEN + 1];

            memset(tapename, 0, sizeof(tapename));
            memcpy(tapename, img_name, img_name_len);
            printf(".. setting tape name to '%s'.\n", tapename);
        }
        /* copy PETSCII tapename into image */
        memcpy(image->tapename, tapename_pet, img_name_len);
    }

    /* report success */
    if (!quiet) {
        printf(".. created new image with %d entries, "
               "%zu ($%zx) bytes.\n",
               index, image->size, image->size);
    }

    return image;
}

/* }}} */
